src/builder.rs

```diff
@@ -0,0 +1,175 @@
+use std::borrow::Cow;
+use { Event, Attribute };
+
+pub struct EventsBuilder<'a> {
+  events: Vec<Event<'a>>,
+}
+
+pub struct EventBuilder<'a> {
+  event: Event<'a>,
+}
+
+pub struct AttributesBuilder<'a> {
+  attrs: Cow<'a, [Attribute<'a>]>,
+}
+
+impl<'a> EventsBuilder<'a> {
+  pub fn new() -> EventsBuilder<'a> {
+    EventsBuilder { events: Vec::new() }
+  }
+
+  pub fn text<T>(mut self, text: T) -> EventsBuilder<'a>
+    where T: Into<Cow<'a, str>>
+  {
+    self.events.push(Event::Text(text.into()));
+    self
+  }
+
+  pub fn tag<N, A, I>(mut self, name: N, attrs: A, inner: I) -> EventsBuilder<'a>
```

Comment:
This is not the correct design for a sequential builder of things. We are "adding" stuff to the "set of events", not modifying a single item. If there was a `TagBuilder`, then it would be useful to have an `attr` fn with the same pattern, since we are modifying a single tag (and potentially a tag that was assigned before).

Using a fn with:
```rust
pub fn put_tag<N, A, I>(&mut self, name: N, attrs: A, inner: I) -> EventsBuilder<'a> ...
```
has an advantage:
```diff
-    pub fn text<T>(mut self, text: T) -> EventsBuilder<'a>
+    pub fn text<T>(&mut self, text: T)
         where T: Into<Cow<'a, str>>
     {
         self.events.push(Event::Text(text.into()));
-        self
     }

-    pub fn tag<N, A, I>(mut self, name: N, attrs: A, inner: I) -> EventsBuilder<'a>
+    pub fn tag<N, A, I>(&mut self, name: N, attrs: A, inner: I)
         where N: Into<Cow<'a, str>>,
               A: Into<AttributesBuilder<'a>>,
-              I: FnOnce(EventsBuilder<'a>) -> EventsBuilder<'a>
+              I: FnOnce(&mut EventsBuilder<'a>)
     {
         let name = name.into();
         self.events.push(Event::StartTag {
             name: name.clone(),
             attrs: attrs.into().attrs,
         });
-        self.events.append(&mut inner(EventsBuilder::new()).into());
+        inner(self);
+        // work on the same self (no copy) <------------------------
         self.events.push(Event::EndTag { name: name });
-        self
     }
 }

@@ -158,21 +156,20 @@ mod tests {

     #[test]
     fn test_prefix() {
-        let events = EventsBuilder::new()
-                         .tag("h1", [("id", "hello")], |events| {
-                             events.text("Hello, ")
-                                   .tag("small", [("id", "world")], |events| events.text("world"))
-                         })
-                         .into();
+        let events = EventsBuilder::new();
+        events.tag("h1", [("id", "hello")], |events| {
+            events.text("Hello, ");
+            events.tag("small", [("id", "world")], |events| events.text("world"))
+        });
```